qiskit_optimization/algorithms/qrao/quantum_random_access_encoding.py

```diff
@@ -0,0 +1,570 @@
+        ValueError: If there are unused decision variables in `x` after mapping to qubits.
+    """
+    # Create a set of all remaining decision variables
+    remaining_dvars = set(range(len(x)))
+    # Create a list to store the binary mappings of each qubit to its corresponding decision variables
+    variable_mappings: List[List[int]] = []
+    # Check that each qubit is associated with at most max_vars_per_qubit variables
+    for qi_vars in q2vars:
+        if len(qi_vars) > max_vars_per_qubit:
+            raise ValueError(
+                "Each qubit is expected to be associated with at most "
+                f"`max_vars_per_qubit` ({max_vars_per_qubit}) variables, "
+                f"not {len(qi_vars)} variables."
+            )
+        # Create a list to store the binary mapping of the current qubit
+        qi_bits: List[int] = []
+
+        # Map each decision variable associated with the current qubit to a binary value and add it
+        # to the qubit bits
+        for dvar in qi_vars:
+            try:
+                qi_bits.append(x[dvar])
+            except IndexError:
+                raise ValueError(f"Decision variable not included in dvars: {dvar}") from None
+            try:
+                remaining_dvars.remove(dvar)
+            except KeyError:
+                raise ValueError(
+                    f"Unused decision variable(s) in dvars: {remaining_dvars}"
+                ) from None
+
+        # Pad with zeros if necessary
+        while len(qi_bits) < max_vars_per_qubit:
+            qi_bits.append(0)
+
+        variable_mappings.append(qi_bits)
+
+    # Raise an error if not all decision variables are used
+    if remaining_dvars:
+        raise ValueError(f"Not all dvars were included in q2vars: {remaining_dvars}")
+
+    # Prepare the individual qrac circuit and combine them into a multi qubit circuit
+    qracs = [_qrac_state_prep_1q(qi_bits) for qi_bits in variable_mappings]
+    qrac_circ = reduce(lambda x, y: x.tensor(y), qracs)
+    return qrac_circ
+
+
+class QuantumRandomAccessEncoding:
+    """This class specifies a Quantum Random Access Code that can be used to encode
+    the binary variables of a QUBO (quadratic unconstrained binary optimization
+    problem).
+
+    Args:
+        max_vars_per_qubit: maximum possible compression ratio.
+            Supported values are 1, 2, or 3.
+
+    """
+
+    # This defines the convention of the Pauli operators (and their ordering)
+    # for each encoding.
+    OPERATORS = (
+        (SparsePauliOp("Z"),),  # (1,1,1) QRAC
+        (SparsePauliOp("X"), SparsePauliOp("Z")),  # (2,1,p) QRAC, p ≈ 0.85
+        (SparsePauliOp("X"), SparsePauliOp("Y"), SparsePauliOp("Z")),  # (3,1,p) QRAC, p ≈ 0.79
+    )
+
+    def __init__(self, max_vars_per_qubit: int = 3):
+        if max_vars_per_qubit not in (1, 2, 3):
+            raise ValueError("max_vars_per_qubit must be 1, 2, or 3")
```

Comment:
Maybe dues to this outer check the ifs that I commented above that assumed this range omitted doing anything further on the basis it was done here. While this guards from the end user point of view I think the ifs I commented on should still check to ensure code correctness etc.